feat(event_handler): support Header parameter validation in OpenAPI schema (#3687)
awslabs/aws-lambda-powertools-python: docs/core/event_handler/api_gateway.md

1. `Path` is a special OpenAPI type that allows us to constrain todo_id to be less than 999.

#### Validating headers

> We use the `Annotated` type to tell Event Handler that a particular parameter is a header that needs to be validated.

Reviewer: ```suggestion
We use the `Annotated` type to tell the Event Handler that a particular parameter is a header that needs to be validated.
```

> !!! info "We adhere to HTTP RFC standards, which means we treat HTTP headers as case-insensitive."

Reviewer: Can we add a link to the RFC standards like you did in the code?

In the following example, we use a new `Header` OpenAPI type to add [one out of many possible constraints](#customizing-openapi-parameters), which should read as:

* `correlation_id` is a header that must be present in the request

> * `correlation_id`, when set, should have 16 characters

Reviewer: ```suggestion
* `correlation_id` should have 16 characters
```

The previous bullet point already makes it required

> * Doesn't match? Event Handler will return a validation error response

Reviewer: ```suggestion
* No match? Event Handler will return a validation error response
```

<!-- markdownlint-disable MD013 -->

=== "validating_headers.py"

    ```python hl_lines="8 10 29"
    --8<-- "examples/event_handler_rest/src/validating_headers.py"
    ```

    1. If you're not using Python 3.9 or higher, you can install and use [`typing_extensions`](https://pypi.org/project/typing-extensions/){target="_blank" rel="nofollow"} to the same effect

>     2. `Header` is a special OpenAPI type that can add constraints to a header well as document them

Reviewer: ```suggestion
    2. `Header` is a special OpenAPI type that can add constraints and documentation to a header
```

>     3. **First time seeing the `Annotated`?** <br><br> This special type uses the first argument as the actual type, and subsequent arguments are metadata. <br><br> At runtime, static checkers will also see the first argument, but anyone receiving them could inspect them to fetch their metadata.

Reviewer: ```suggestion
    3. **First time seeing `Annotated`?** <br><br> This special type uses the first argument as the actual type, and subsequent arguments as metadata. <br><br> At runtime, static checkers will also see the first argument, but any receiver can inspect it to get the metadata.
```

=== "working_with_headers_multi_value.py"

>     If you need to handle multi-value for specific headers, you can create a list of the desired type.

Reviewer: ```suggestion
    You can handle multi-value headers by declaring it as a list of the desired type.
```
